OpenSim/Common/DataAdapter.h

```diff
@@ -0,0 +1,225 @@
+ * Concrete classes handle the details (e.g. format, sequential access, etc...) 
+ * associated with a particular data source/sink.
+ *
+ * The base DataAdapter contains a static registry to serve as a factory for 
+ * concrete DataAdpaters, given a string identifier of the type of adapter.
+ * The adapter knows the source format and data flow (read, write, both).
+ * String identifiers can be associated with file formats according to known
+ * file extensions.
+ *
+ * @author Ajay Seth
+ */
+class OSIMCOMMON_API DataAdapter { 
+//=============================================================================
+// METHODS
+//=============================================================================
+public:
+    virtual ~DataAdapter(){};
+
+    virtual DataAdapter* clone() const = 0;
+
+    /** Creator of concrete DataAdapter(s) for the specified source type by its
+        unique identifier (string). For example, for file based sources, a 
+        component can acquire the necessary adapter instance to read the data
+        by the file extension if the extension is used as its identifier.
+    @param[in] identifier string used to uniquely identify the concrete adapter
+               that processes data from/to sources/sinks of a particular type.
+               For file adapters, .*/
+    static DataAdapter* createAdapter(const std::string& identifier);
+
+    /** Register a concrete DataAdapter by its unique string identifier.
+        Registration permits access to the required concrete adapter by
+        identifier lookup. As such, identifiers must be unique, but adapters may
+        be registered with multiple identifiers. For example, a data file may 
+        have multiple valid extensions (e.g. ".jpg: and ".jpeg") in which case
+        both extensions would be valid identifiers for the same adapter. If an
+        identifier is already in use an Exception is thrown. 
+    @param[in] identifier string used to uniquely identify the required
+               concrete adapter to process a source/sink. For file adapters,
+               register unique file extension(s) in order for OpenSim to read/
+               write different file formats.
+    @param[in] adapter the concrete DataAdapter required to process the type
+                       (format) of the data for the source/sink.*/
+    static void registerDataAdapter(const std::string& identifier,
+        const DataAdapter& adapter);
+
+    /** Convenience construction of a DataTable from a data source identified
+        by name and its type.
+    @param[in] source           string identifying the source to be opened
+    @param[in] sourceTypeName   string identifying the source type used to
+                                determine which concrete adapter is used.
+    @return DataTable           the resulting in memory data from the source
+        */
+    AbstractDataTable* createDataTable(const std::string& source,
```

Comment:
It shoudn't be static (i don't see it marked as such) and I should have removed the method altogether.  I was toying with the idea of getting a table from the adapter. The table already has facilities for using an adapter to read from a file where it extracts the type from the extension. But that is still a special case for convenience in my book. I am not saying this is at all reasonable or better, but in general we do want to be able to talk to a device (port) and establish which adapter to process the contents into a table. The first argument is meant to identify the source and it could be a port address and the other is to identify the adapter that could intepret the data coming through it. I realize using strings as argument types might not make sense. 

I would prefer not to give a DataTable any information about where data is coming from (a source) including the file case, but it did seem awfully convenient so I left it in there. I think only the concrete adapter should know about files and other potential sources of data.